azure-kusto-ingest/tests/test_storage_account_set.py

```diff
@@ -0,0 +1,97 @@
+        # Use set to verify same accounts are in both lists
+        a = set(ranked_accounts_1)
+        b = set(ranked_accounts_2)
+        self.assertEqual(a, b)
+        # Use != to verify that the lists are not identical in order
+        self.assertNotEqual(ranked_accounts_1, ranked_accounts_2)
+
+    def test_check_rank_when_all_failure(self):
+        current_time = 0
+
+        def time_provider():
+            return current_time
+
+        storage_account_set = self.create_storage_account_set(time_provider)
+
+        # Simulate data in 10 seconds passing
+        for current_time in range(10):
+            storage_account_set.add_account_result(self.ACCOUNT_1, False)
+            storage_account_set.add_account_result(self.ACCOUNT_2, False)
+            storage_account_set.add_account_result(self.ACCOUNT_3, False)
+            storage_account_set.add_account_result(self.ACCOUNT_4, False)
+            storage_account_set.add_account_result(self.ACCOUNT_5, False)
+
+        # All accounts should have the same rank (0)
+        self.assertEqual(storage_account_set.get_storage_account(self.ACCOUNT_1).get_rank(), 0)
+        self.assertEqual(storage_account_set.get_storage_account(self.ACCOUNT_2).get_rank(), 0)
+        self.assertEqual(storage_account_set.get_storage_account(self.ACCOUNT_3).get_rank(), 0)
+        self.assertEqual(storage_account_set.get_storage_account(self.ACCOUNT_4).get_rank(), 0)
+        self.assertEqual(storage_account_set.get_storage_account(self.ACCOUNT_5).get_rank(), 0)
+
+    def test_check_rank_when_success_rate_is_different(self):
+        current_time = 0
+
+        def time_provider():
+            return current_time
+
+        storage_account_set = self.create_storage_account_set(time_provider)
+
+        # Simulate data in 30 seconds passing
+        for current_time in range(60):
+            storage_account_set.add_account_result(self.ACCOUNT_1, True)  # 100% success
+            storage_account_set.add_account_result(self.ACCOUNT_2, current_time % 10 != 0)  # ~90% success
+            storage_account_set.add_account_result(self.ACCOUNT_4, current_time % 2 == 0)  # ~40% success
+            storage_account_set.add_account_result(self.ACCOUNT_3, current_time % 3 == 0)  # ~33% success
+            storage_account_set.add_account_result(self.ACCOUNT_5, False)  # 0% success
+
+        # Get shuffled accounts
+        ranked_accounts = storage_account_set.get_ranked_shuffled_accounts()
+        # Verify that the accounts are ranked in the correct order
+        self.assertIn(ranked_accounts[0].get_account_name(), [self.ACCOUNT_1, self.ACCOUNT_2])
+        self.assertIn(ranked_accounts[1].get_account_name(), [self.ACCOUNT_1, self.ACCOUNT_2])
+        self.assertIn(ranked_accounts[2].get_account_name(), [self.ACCOUNT_3, self.ACCOUNT_4])
+        self.assertIn(ranked_accounts[3].get_account_name(), [self.ACCOUNT_3, self.ACCOUNT_4])
+        self.assertEqual(ranked_accounts[4].get_account_name(), self.ACCOUNT_5)
```

Comment:
Something seemed wrong with your algorithm, I'm not sure what it is though.
I tested it, and the results don't seem to be right.
I'll give you the tests to check yourself and tell me what you think.
The assert == 0.5 is a placeholder, it should be a different number by the buckets ratio, but it still seems wrong.

```python
    def test_multiple_results(self):
        current_time = 0

        def time_provider():
            return current_time

        storage_account_set = self.create_storage_account_set(time_provider)

        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        storage_account_set.add_account_result(self.ACCOUNT_1, False)

        assert storage_account_set.accounts[self.ACCOUNT_1].get_rank() == 0.5

    def test_full_buckets(self):
        current_time = 0

        def time_provider():
            return current_time

        storage_account_set = self.create_storage_account_set(time_provider)

        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11

        assert storage_account_set.accounts[self.ACCOUNT_1].get_rank() == 0


    def test_buckets_are_recycled(self):
        current_time = 0

        def time_provider():
            return current_time

        storage_account_set = self.create_storage_account_set(time_provider)

        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)

        assert storage_account_set.accounts[self.ACCOUNT_1].get_rank() == 0

    def test_small_deltas(self):
        current_time = 0

        def time_provider():
            return current_time
        storage_account_set = self.create_storage_account_set(time_provider)

        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 3
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 3
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 3
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 3
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 3
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 3

        assert storage_account_set.accounts[self.ACCOUNT_1].get_rank() == 0.5

    def test_big_break(self):
        current_time = 0

        def time_provider():
            return current_time

        storage_account_set = self.create_storage_account_set(time_provider)
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 90
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, True)
        current_time += 11
        storage_account_set.add_account_result(self.ACCOUNT_1, False)
        current_time += 11

        assert storage_account_set.accounts[self.ACCOUNT_1].get_rank() == 0.5

```